Add Support to Configure Leaf Certificate Validity Period in Karmada Operator

jabellard opened the pull request:
**What type of PR is this?**

/kind feature

<!--
Add one of the following kinds:

/kind api-change
/kind bug
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind feature
/kind flake

-->

**What this PR does / why we need it**:
This is an implementation of this [proposal](https://github.com/karmada-io/karmada/pull/6176) to add support to the Karmada operator for configuring the validity period of control plane leaf certificates.

### With Default Validity
```shell
mc -n karmada-system get karmadas.operator.karmada.io t1 -oyaml | yq '.spec.customCertificate'
```

```yaml
null
```

```shell
mc -n karmada-system get secrets t1-karmada-cert -o jsonpath='{.data.apiserver\.crt}' | base64 --decode | openssl x509 -text -noout | grep -i "not after"
```

```text
Not After : Mar  9 16:10:09 2026 GMT
```

### With Custom Validity

```shell
mc -n karmada-system get karmadas.operator.karmada.io t2 -oyaml | yq '.spec.customCertificate'
```

```yaml
leafCertValidityDays: 90
```

```shell
mc -n karmada-system get secrets t2-karmada-cert -o jsonpath='{.data.apiserver\.crt}' | base64 --decode | openssl x509 -text -noout | grep -i "not after"
```

```text
Not After : Jun  7 16:12:19 2025 GMT
```


**Which issue(s) this PR fixes**:
Part of #6174
**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required.
-->
```release-note
`karmada-operator`: The new `spec.customCertificate.leafCertValidityDays` field can optionally be used to specify a custom validity period in days for control plane leaf certificates (e.g., API server certificate). When not explicitly set, the default validity period of 1 year is used.
```

RainbowMango commented:
Hi Joe, I added some other iterated tasks to #6174, such as documentation and tests. I also changed the PR description a little bit (fixes --> part of)to avoid this PR closing the original issue.

jabellard commented:
> Hi Joe, I added some other iterated tasks to #6174, such as documentation and tests. I also changed the PR description a little bit (fixes --> part of)to avoid this PR closing the original issue.

Ok.

zhzhuang-zju commented:
/assign

zhzhuang-zju commented on operator/pkg/apis/operator/v1alpha1/type.go:
```diff
@@ -133,6 +133,11 @@ type CustomCertificate struct {
+	// LeafCertValidityDays specifies the validity period of leaf certificates (e.g., API Server certificate) in days.
+	// If not specified, the default validity period of 1 year will be used.
+	// +optional
+	LeafCertValidityDays *int64 `json:"leafCertValidityDays,omitempty"`
```
Do we need to determine whether `LeafCertValidityDays` is greater than 0?

jabellard replied:
Don't think that's needed. Type is nullable.

jabellard replied:
Updated to use 32 bit int

zhzhuang-zju replied:
The generated certificate will be unavailable when `leafCertVatlidityDays` is lower than 0.
```bash
$ kubectl apply -v=4 -f - <<EOF
apiVersion: operator.karmada.io/v1alpha1
kind: Karmada
metadata:
  name: karmada-demo
  namespace: test
spec:
  customCertificate:
    leafCertValidityDays: -365
  components:
    karmadaDescheduler: {}
    etcd: {}
EOF
$ kubectl logs -p --namespace test karmada-demo-apiserver-67b9fbf674-zd8dr
W0312 01:49:19.615519       1 logging.go:55] [core] [Channel #1 SubChannel #4]grpc: addrConn.createTransport failed to connect to {Addr: "karmada-demo-etcd-client.test.svc.cluster.local:2379", ServerName: "karmada-demo-etcd-client.test.svc.cluster.local:2379", }. Err: connection error: desc = "transport: authentication handshake failed: tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2025-03-12T01:49:19Z is after 2024-03-12T01:47:57Z"
```

jabellard replied:
Just pushed changes to address this

zhzhuang-zju commented on operator/pkg/tasks/init/cert.go:
```diff
@@ -85,6 +86,11 @@ func newCertSubTasks(karmada *operatorv1alpha1.Karmada) []workflow.Task {
 			task = workflow.Task{Name: cert.Name, Run: runCATask(cert)}
 			caCert[cert.Name] = cert
 		} else {
+			if karmada.Spec.CustomCertificate != nil && karmada.Spec.CustomCertificate.LeafCertValidityDays != nil {
```
It is recommended to move the places where `LeafCertValidityDays` is applied to the [`function mutateCertConfig`](https://github.com/karmada-io/karmada/blob/d6cc0c22db94e7104daae39fbd9571fcdba3349e/operator/pkg/tasks/init/cert.go#L189). Like
```go
func mutateCertConfig(data InitData, cc *certs.CertConfig) error {
	if cc.AltNamesMutatorFunc != nil {
		err := cc.AltNamesMutatorFunc(&certs.AltNamesMutatorConfig{
			Name:                data.GetName(),
			Namespace:           data.GetNamespace(),
			Components:          data.Components(),
			ControlplaneAddress: data.ControlplaneAddress(),
		}, cc)

		if err != nil {
			return err
		}
	}

	if data.CustomCertificate().LeafCertValidityDays != nil {
		certValidityDuration := time.Hour * 24 * time.Duration(*data.CustomCertificate().LeafCertValidityDays)
		notAfter := time.Now().Add(certValidityDuration).UTC()
		cc.NotAfter = &notAfter
	}

	return nil
}
```



The benefits of doing so are as follows: 
- The modifications are more concentrated.
- Easier to expand. For example, other certificate configurations in the future can all be implemented in the function `mutateCertConfig`.

jabellard replied:
Good point. Updated.

zhzhuang-zju commented:
thanks
/lgtm

RainbowMango reviewed (commented):
/assign

RainbowMango reviewed (commented):
/lgtm

RainbowMango commented:
/approve

jabellard commented:
@RainbowMango , can this feature be included as part of the next alpha release? I want to use it.

RainbowMango commented:
Yes, I think so, as long as all the PRs you need have been merged by the end of this week. Please let me know if you need more PRs related to this feature.

By the way, the next alpha will be cut from the master as per the [Release Timeline](https://karmada.io/docs/releases#release-timeline) by the end of this month(probably by the end of this week because of the KubeCon).

jabellard commented:
> Yes, I think so, as long as all the PRs you need have been merged by the end of this week. Please let me know if you need more PRs related to this feature.
> 
> By the way, the next alpha will be cut from the master as per the [Release Timeline](https://karmada.io/docs/releases#release-timeline) by the end of this month(probably by the end of this week because of the KubeCon).

Yup. This should be ready to go. Thanks.
